File: src/main/java/me/modmuss50/fr/PipeModel.java
package me.modmuss50.fr;

import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.*;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.resources.model.IBakedModel;
import net.minecraft.client.resources.model.ModelRotation;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.client.model.ISmartBlockModel;
import org.lwjgl.util.vector.Vector3f;
import reborncore.common.misc.vecmath.Vecs3dCube;

import java.util.ArrayList;
import java.util.List;

public class PipeModel implements ISmartBlockModel {

    static FaceBakery faceBakery = new FaceBakery();
    TextureAtlasSprite texture = null;

    public PipeModel() {
        texture = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite("minecraft:blocks/stone");
    }

    public PipeModel(IBlockState blockState) {
        String textureName = blockState.getValue(BlockPipe.TYPE_PROP).textureName;
        texture = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(textureName);
    }

    @Override
    public IBakedModel handleBlockState(IBlockState state) {
        return new PipeModel(state);
    }

    @Override
    public List<BakedQuad> getFaceQuads(EnumFacing p_177551_1_) {
        return new ArrayList<>();
    }

    @Override
    public List<BakedQuad> getGeneralQuads() {
        ArrayList<BakedQuad> list = new ArrayList();
        BlockFaceUV uv = new BlockFaceUV(new float[]{0.0F, 0.0F, 16.0F, 16.0F}, 0);
        BlockPartFace face = new BlockPartFace(null, 0, "", uv);
        addCubeToList(new Vecs3dCube(4, 4, 4, 12, 12, 12), list, face, ModelRotation.X0_Y0);
        return list;
    }

    public void addCubeToList(Vecs3dCube cube, ArrayList<BakedQuad> list, BlockPartFace face, ModelRotation modelRotation){
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMinX(), (float)cube.getMinY(), (float)cube.getMinZ()), new Vector3f((float)cube.getMaxX(), (float)cube.getMinY(), (float)cube.getMaxZ()), face, texture, EnumFacing.DOWN, modelRotation, null, true, true));//down
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMinX(), (float)cube.getMaxY(), (float)cube.getMinZ()), new Vector3f((float)cube.getMaxX(), (float)cube.getMaxY(), (float)cube.getMaxZ()), face, texture, EnumFacing.UP, modelRotation, null, true, true));//up
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMinX(), (float)cube.getMinY(), (float)cube.getMinZ()), new Vector3f((float)cube.getMaxX(), (float)cube.getMaxY(), (float)cube.getMaxZ()), face, texture, EnumFacing.NORTH, modelRotation, null, true, true));//north
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMinX(), (float)cube.getMinY(), (float)cube.getMaxZ()), new Vector3f((float)cube.getMaxZ(), (float)cube.getMaxY(), (float)cube.getMaxZ()), face, texture, EnumFacing.SOUTH, modelRotation, null, true, true));//south
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMaxX(), (float)cube.getMinY(), (float)cube.getMinZ()), new Vector3f((float)cube.getMaxX(), (float)cube.getMaxY(), (float)cube.getMaxZ()), face, texture, EnumFacing.EAST, modelRotation, null, true, true));//east
        list.add(faceBakery.makeBakedQuad(new Vector3f((float)cube.getMinX(), (float)cube.getMinY(), (float)cube.getMinZ()), new Vector3f((float)cube.getMinX(), (float)cube.getMaxY(), (float)cube.getMaxZ()), face, texture, EnumFacing.WEST, modelRotation, null, true, true));//west
    }


    @Override
    public boolean isAmbientOcclusion() {
        return false;
    }

    @Override
    public boolean isGui3d() {
        return true;
    }

    @Override
    public boolean isBuiltInRenderer() {
        return false;
    }

    @Override
    public TextureAtlasSprite getParticleTexture() {
        return texture;
    }

    @Override
    public ItemCameraTransforms getItemCameraTransforms() {
        return null;
    }
}

File: src/main/kotlin/me/modmuss50/fr/client/PipeModel.kt
package me.modmuss50.fr.client

import com.google.common.collect.ImmutableMap
import me.modmuss50.fr.FluxedRedstone
import me.modmuss50.fr.mutlipart.PipeMultipart
import me.modmuss50.fr.mutlipart.PipeTypeEnum
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.*
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.util.EnumFacing
import net.minecraftforge.client.model.IPerspectiveAwareModel
import net.minecraftforge.common.model.TRSRTransformation
import net.minecraftforge.common.property.IExtendedBlockState
import org.apache.commons.lang3.tuple.Pair
import reborncore.client.models.BakedModelUtils
import reborncore.common.misc.vecmath.Vecs3dCube
import java.util.*
import javax.vecmath.Matrix4f


class PipeModel(var type: PipeTypeEnum, var transforms: ImmutableMap<ItemCameraTransforms.TransformType, TRSRTransformation>) : IBakedModel, IPerspectiveAwareModel {


    internal var faceBakery = FaceBakery()
    internal var texture: TextureAtlasSprite? = null

    init {
        texture = Minecraft.getMinecraft().textureMapBlocks.getAtlasSprite(type.textureName)
    }

    override fun getQuads(blockState: IBlockState?, p1: EnumFacing?, p2: Long): MutableList<BakedQuad>? {
        if (blockState != null) {
            type = blockState.getValue(PipeMultipart.TYPE)
            texture = Minecraft.getMinecraft().textureMapBlocks.getAtlasSprite(type.textureName)
        }
        val list = ArrayList<BakedQuad>()
        val thickness = type.thickness!!
        val lastThickness = 16 - thickness
        val uv = BlockFaceUV(floatArrayOf(0.0f, 0.0f, 16.0f, 16.0f), 0)
        val face = BlockPartFace(null, 0, "", uv)
        BakedModelUtils.addCubeToList(Vecs3dCube(thickness, thickness, thickness, lastThickness, lastThickness, lastThickness), list, face, ModelRotation.X0_Y0, texture!!, EnumFacing.DOWN, faceBakery)
        if (blockState is IExtendedBlockState && p1 != null) {
            val state = blockState
            if (state.getValue(PipeMultipart.UP)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(thickness, lastThickness, thickness, lastThickness, 16.0, lastThickness), list, face, ModelRotation.X0_Y0, texture!!, p1, faceBakery)
            }
            if (state.getValue(PipeMultipart.DOWN)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(thickness, 0.0, thickness, lastThickness, thickness, lastThickness), list, face, ModelRotation.X0_Y0, texture!!, p1, faceBakery)
            }
            if (state.getValue(PipeMultipart.NORTH)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(0.0, thickness, thickness, thickness, lastThickness, lastThickness), list, face, ModelRotation.X0_Y90, texture!!, p1, faceBakery)
            }
            if (state.getValue(PipeMultipart.SOUTH)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(0.0, thickness, thickness, thickness, lastThickness, lastThickness), list, face, ModelRotation.X0_Y270, texture!!, p1, faceBakery)
            }
            if (state.getValue(PipeMultipart.EAST)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(lastThickness, thickness, thickness, 16.0, lastThickness, lastThickness), list, face, ModelRotation.X0_Y0, texture!!, p1, faceBakery)
            }
            if (state.getValue(PipeMultipart.WEST)) {
                BakedModelUtils.addCubeToList(Vecs3dCube(0.0, thickness, thickness, thickness, lastThickness, lastThickness), list, face, ModelRotation.X0_Y0, texture!!, p1, faceBakery)
            }
        }
        return list
    }

    override fun isAmbientOcclusion(): Boolean {
        return false
    }

    override fun isGui3d(): Boolean {
        return true
    }

    override fun isBuiltInRenderer(): Boolean {
        return false
    }

    override fun getParticleTexture(): TextureAtlasSprite {
        return texture!!
    }

    override fun getItemCameraTransforms(): ItemCameraTransforms? {
        return ItemCameraTransforms.DEFAULT
    }

    override fun getOverrides(): ItemOverrideList? {
        return ItemOverrideList.NONE
    }

    override fun handlePerspective(cameraTransformType: ItemCameraTransforms.TransformType?): Pair<out IBakedModel, Matrix4f>? {
        val pair = IPerspectiveAwareModel.MapWrapper.handlePerspective(this, transforms, cameraTransformType)
        return Pair.of<PipeModel, Matrix4f>(this, pair.getRight())
    }
}